core/src/transport/circuit_breaking.rs

```diff
@@ -80,37 +85,80 @@ impl CircuitBreakerState {
 
     /// Returns `None` if circuit of `addr` is closed.
     fn get_wait_duration(&self, addr: &Multiaddr) -> Option<Duration> {
+        let circuit_addr = CircuitAddr::from(addr.clone());
         self.addresses
             .lock()
-            .get(addr)
+            .get(&circuit_addr)
             .map(|addr_info| addr_info.wait_duration)
     }
 
     fn open_or_extend_circuit(&self, addr: Multiaddr) {
-        match self.addresses.lock().entry(addr) {
+        let circuit_addr = CircuitAddr::from(addr);
+        match self.addresses.lock().entry(circuit_addr) {
             Entry::Vacant(entry) => {
-                entry.insert(AddrInfo {
+                let inserted = entry.insert(AddrInfo {
                     wait_duration: INITIAL_TIMEOUT,
-                    expiration: Delay::new(INITIAL_TIMEOUT * 2),
+                    expiration: Instant::now() + INITIAL_TIMEOUT * 2,
                 });
+                debug!("Opened circuit: {:?}", inserted);
             }
 
             Entry::Occupied(entry) => {
+                let addr = entry.key().clone();
                 let addr_info = entry.into_mut();
                 addr_info.wait_duration = cmp::min(addr_info.wait_duration * 2, MAX_TIMEOUT);
-                addr_info.expiration = Delay::new(addr_info.wait_duration * 2);
+                addr_info.expiration = Instant::now() + addr_info.wait_duration * 2;
+                debug!("Extended circuit {:?}: {:?}", addr, addr_info);
             }
         }
     }
 
     /// Closes the circuit of `addr` even if the circuit has not yet expired.
     fn close_circuit(&self, addr: &Multiaddr) {
-        self.addresses.lock().remove(addr);
+        let circuit_addr = CircuitAddr::from(addr.clone());
+        let removed = self.addresses.lock().remove(&circuit_addr);
+        debug!("Closed circuit {:?}", removed);
     }
 
     fn close_expired_circuits(&self) {
+        let now = Instant::now();
         let mut addresses = self.addresses.lock();
-        addresses.retain(move |_, a| (&mut a.expiration).now_or_never().is_none());
+        addresses.retain(|_, a| a.expiration > now);
+        debug!("Closed expired circuits")
+    }
+}
+
+/// Wraps a `Multiaddr` whose `PeerId` (if any) has been
+/// [decapsulated](https://github.com/libp2p/specs/blob/master/addressing/README.md#decapsulation).
+///
+/// In large public networks (e.g. IPFS) many nodes advertise LAN local
+/// addresses. In most cases nodes don't share the same LAN and thus dialing
+/// these addresses fails.
+///
+/// Ignoring the `PeerId` in addresses prevents dialing e.g. `192.168.2.108`
+/// multiple times in a row.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+struct CircuitAddr(Multiaddr);
+
+impl From<Multiaddr> for CircuitAddr {
+    fn from(mut addr: Multiaddr) -> Self {
+        // Check if `addr` contains `Protocol::P2p(_)`.
+        let contained_p2p = addr.iter().find(|protocol| match protocol {
```

Comment:
```suggestion
        let contained_p2p = addr.iter().any(|protocol| matches!(protocol, Protocol::P2p(_));
```

Might a bit simpler, right?

Reply:
True :) Addressed in f64e6aa.